### DISCUSSION TITLE
Energy scan without parameter report a latched RSSI value

### DISCUSSION THREAD

**Discussion by maeleph** [openthread/openthread#8713] [Q&A] [answered]
Hello, 

Current implementation of the energy scan without parameter function reports a latched RSSI value. This might not be representative of the actual energy level on a given channel, as in most of the cases, we have never received any frame on RF channels other than the active network channel.

An alternative could be to report the instantaneous RSSI level. However, most of the time, we would not capture actual traffic on the channel as we might probe in between frames transmission. So this might also not be representative of the actual activity on a given RF channel. 

Another option could be to use the energy scan with a parameter to measure the energy level of the RF channel during a given time window. This would provide more meaningful results, but this would have some drawbacks as it will increase the time during which another RF channel is selected and as a consequence the Thread device is not able to receive frames. 

Any thoughts on how to improve the current channel monitoring feature ?

**Comment by jwhui** [comment, marked as answer]
You may find [OpenThread's Channel Monitoring](https://openthread.io/reference/group/api-channel-monitor) feature useful.

> This might not be representative of the actual energy level on a given channel, as in most of the cases, we have never received any frame on RF channels other than the active network channel.

Note that failure to receive frames is not an indication that a channel is clean. There may be other forms of 2.4 GHz interference that are not related to IEEE 802.15.4.

**Reply by maeleph** [reply]
Thanks for the reply. 

Our question was mostly that we think that the channel monitoring cannot work properly as implemented currently. 

As there are no new RSSI measurement done during the channel monitoring procedure, we entirely base the results of this on either : 
- old measures : we could have received a message 1ms ago or 1 minute ago or more
- non-existent results : if we never switch channel, there can be no RSSI measurements stored for this channel
So the results of the channel monitoring does not necessarily represent the actual state of the network.

**Reply by jwhui** [reply]
[Energy scan](https://openthread.io/reference/group/radio-operation#otplatradioenergyscan) should provide a measure of RF energy regardless of whether a packet was received or not. Note that energy scan implementation is provided by the platform. It is up to the platform drivers to implement energy scan properly.

**Reply by maeleph** [reply]
As per description of the channel monitoring : 

> When channel monitoring is active, a zero-duration Energy Scan is performed, collecting a single RSSI sample on every channel per sample interval. 

In the code we call the function GetRssi() which description is : 

> 
> /**
>      * This method gets the most recent RSSI measurement from radio link.
>      *
>      * @returns The RSSI in dBm when it is valid. `kInvalidRssiValue` when RSSI is invalid.
>      *
>      */

Our understanding of this is that we should not trigger a new measurement, but return the last value that we had measured.
Is that correct ?

**Reply by jwhui** [reply]
Radios typically maintain a running RSSI average over a sliding window (often over the last 8 symbols) while the receiver is enabled. Every symbol time, a new average is latched.

**Reply by maeleph** [reply]
in case of RCP, the host is supposed to select channel, and trigger measurement. How would this work, should the radio switch channel to make measurements on other channels ?

**Reply by jwhui** [reply]
Yes, the radio must switch to the channel requested for the energy scan.

**Reply by abtink** [reply]
The "channel monitoring" feature and energy scan have been both validated and tested on devices and work as intended.

As @jwhui  mentioned, radiio platform should return the recent RSSI measurement from `otPlatRadioGetRssi()` API  (taken on the current rx channel). I think the details on how the RSSI sample is taken (e.g., averaging over last 8 received symbols, etc) are specified by 802.15.4 standard.

RCP looks like any radio platform from perspective of host code.

Energy Scan implementation does indeed switch the channel.
In particular for zero-duration energy scan: we put radio in rx on the scan channel, take an RSSI sample measurement, switch to next channel (in the given channel scan mask) if any:
https://github.com/openthread/openthread/blob/48128acb76b24713158096ebd64932822d5272aa/src/core/mac/mac.cpp#L312-L345
